# 머신러닝 with Python 
## 09 앙상블 학습 그래디언트 부스팅(Boosting) 실습- breast_cancer 데이터

##### 그래디언트 부스팅 알고리즘을 활용해 유방암 여부를 예측하는 모형을 만들어본다.

책 p.286~ 

---

기본 실습은 사이킷런으로 진행하지만 하단부에서는 XGBoost와 LightGBM을 데려와서 실습 해보기도 해보고싶다. 아마 다른 파일에서 진행할 것 같다.

> 배운점1: 

> 배운점2: 

### 데이터 불러오기

In [1]:
from sklearn import datasets

In [2]:
# 유방암 데이터 가져오기

raw_breast_cancer = datasets.load_breast_cancer()

In [3]:
# 데이터 살펴보기

raw_breast_cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [4]:
# 데이터 셋 내 피처 이름들

raw_breast_cancer.feature_names

array(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error',
       'fractal dimension error', 'worst radius', 'worst texture',
       'worst perimeter', 'worst area', 'worst smoothness',
       'worst compactness', 'worst concavity', 'worst concave points',
       'worst symmetry', 'worst fractal dimension'], dtype='<U23')

### 피처, 타깃 데이터 지정

In [5]:
X = raw_breast_cancer.data
y = raw_breast_cancer.target

### 트레이닝/테스트 데이터 분할

In [6]:
from sklearn.model_selection import train_test_split                 # 분할을 위해 필요한 함수
X_tn, X_te, y_tn, y_te = train_test_split(X, y, random_state = 0)    # 분리. randomstate는 고정

### 데이터 표준화

In [7]:
from sklearn.preprocessing import StandardScaler         # 데이터 표준화를 위한 함수
std_scale = StandardScaler()                             # 표준화 스케일러 지정
std_scale.fit(X_tn)                                      # 트레이닝 피처를 기준으로 표준화를 적합

X_tn_std = std_scale.transform(X_tn) 
X_te_std = std_scale.transform(X_te)                     # 트레인, 테스트 데이터 각각 적합시킨 표준화에 맞게 변형

### 데이터 학습

In [10]:
from sklearn.ensemble import GradientBoostingClassifier     # 그래디언트 부스팅 함수. 회귀문제를 풀때는 regressor로 호출.
clf_gbt = GradientBoostingClassifier(max_depth = 2,         # 개별 학습기의 최대 깊이
                                     learning_rate = 0.01,  # 학습율. 너무 크면 튀고 너무 작으면 느리다.
                                     random_state = 0)
clf_gbt.fit(X_tn_std, y_tn)                                 # 적합

GradientBoostingClassifier(learning_rate=0.01, max_depth=2, random_state=0)

```
max_depth : The maximum depth of the individual regression estimators. The maximum
    depth limits the number of nodes in the tree. Tune this parameter
    for best performance; the best value depends on the interaction
    of the input variables.
```

맥스 뎁스는 default = 3이지만 2ㅗㄹ 바꾼다.

### 데이터 예측

In [11]:
pred_gbt = clf_gbt.predict(X_te_std)                     # std된 테스트 피처 데이터를 넣고 실행하여 결과 확인 
print(pred_gbt)

[0 1 1 1 1 1 1 1 1 1 1 1 1 0 1 0 1 0 0 0 0 0 1 1 0 1 1 0 1 0 1 0 1 0 1 0 1
 0 1 0 1 1 1 1 1 0 1 1 1 0 0 0 0 1 1 1 1 1 1 0 0 0 1 1 0 1 0 0 0 1 1 0 1 1
 0 1 1 1 1 1 0 0 0 1 0 1 1 1 0 0 1 0 1 0 1 1 0 1 1 1 1 1 1 1 0 1 0 1 0 0 1
 0 0 1 1 1 1 1 1 1 1 1 0 1 0 1 1 1 1 1 0 1 1 1 1 1 1 1 0 1 1 1 0]


### 정확도 평가

In [12]:
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_te, pred_gbt)                # 실제값과 예측값을 넣음
print(accuracy)

0.965034965034965


### confusion matrix 확인

In [13]:
from sklearn.metrics import confusion_matrix
conf_matrix = confusion_matrix(y_te, pred_gbt)
print(conf_matrix)

[[49  4]
 [ 1 89]]


### 분류 리포트 확인

In [14]:
from sklearn.metrics import classification_report
class_report = classification_report(y_te, pred_gbt)
print(class_report)

              precision    recall  f1-score   support

           0       0.98      0.92      0.95        53
           1       0.96      0.99      0.97        90

    accuracy                           0.97       143
   macro avg       0.97      0.96      0.96       143
weighted avg       0.97      0.97      0.96       143



### 결과

0.9650이 나쁜 성능은 아니지만 좋은 성능은 아닌 것 같으니 여기서 멈출수 없지, 일부를 바꾸는 연습을 해본다.

---


# 실험

## 실험1: max_depth(default = 3) 조정

In [17]:
from sklearn.ensemble import GradientBoostingClassifier  # 그래디언트 부스팅 모델 함수. 회귀문제를 풀때는 regressor 호출
from sklearn.metrics import accuracy_score               # 정확도 테스트 함수
from sklearn.metrics import classification_report        # 분류 리포트 확인을 위한 함수

for i in range(1, 6):
    clf_gbt = GradientBoostingClassifier(max_depth = i,         # 개별 학습기의 최대 깊이
                                         learning_rate = 0.01,  # 학습율. 너무 크면 튀고 너무 작으면 느리다.
                                         random_state = 0)
    # 데이터 학습
    clf_gbt.fit(X_tn_std, y_tn)
    
    # 데이터 예측
    pred_gbt = clf_gbt.predict(X_te_std) 
    
    # 데이터 정확도 평가
    accuracy = accuracy_score(y_te, pred_gbt)
    
    # 분류 리포트 확인
    class_report = classification_report(y_te, pred_gbt)
    
    # 모델 상황 확인
    
    print("분류기 개별 max_depth: ", i )
    print("모델 정확도: ", accuracy)
    print("모델 분류 리포트")
    print(class_report)
    print("========================================================")

분류기 개별 max_depth:  1
모델 정확도:  0.958041958041958
모델 분류 리포트
              precision    recall  f1-score   support

           0       1.00      0.89      0.94        53
           1       0.94      1.00      0.97        90

    accuracy                           0.96       143
   macro avg       0.97      0.94      0.95       143
weighted avg       0.96      0.96      0.96       143

분류기 개별 max_depth:  2
모델 정확도:  0.965034965034965
모델 분류 리포트
              precision    recall  f1-score   support

           0       0.98      0.92      0.95        53
           1       0.96      0.99      0.97        90

    accuracy                           0.97       143
   macro avg       0.97      0.96      0.96       143
weighted avg       0.97      0.97      0.96       143

분류기 개별 max_depth:  3
모델 정확도:  0.965034965034965
모델 분류 리포트
              precision    recall  f1-score   support

           0       0.94      0.96      0.95        53
           1       0.98      0.97      0.97        90

    accu

### 결과

특정 개수 이상 늘어나면 오히려 성능이 떨어졌다. 예견된 일이기도 했는데, 튜닝과 관련해서는 놓칠 수 없는 부분이기 때문에 실험해 보았다.

학습기의 개수는 어느 선 이상에서 더이상 성능이 상승하지 않을 것으로 예상되는데, 궁금하니 여러개를 해보기로 해본다.

## 실험2: n_estimators(default=100) 조정

In [19]:
from sklearn.ensemble import GradientBoostingClassifier  # 그래디언트 부스팅 모델 함수. 회귀문제를 풀때는 regressor 호출
from sklearn.metrics import accuracy_score               # 정확도 테스트 함수
from sklearn.metrics import classification_report        # 분류 리포트 확인을 위한 함수

n_esti = [10, 25, 50, 75, 90, 100, 120, 150]

for i in n_esti:
    clf_gbt = GradientBoostingClassifier(max_depth = 2,         # 개별 학습기의 최대 깊이
                                         n_estimators = i,
                                         learning_rate = 0.01,  # 학습율. 너무 크면 튀고 너무 작으면 느리다.
                                         random_state = 0)
    # 데이터 학습
    clf_gbt.fit(X_tn_std, y_tn)
    
    # 데이터 예측
    pred_gbt = clf_gbt.predict(X_te_std) 
    
    # 데이터 정확도 평가
    accuracy = accuracy_score(y_te, pred_gbt)
    
    # 분류 리포트 확인
    class_report = classification_report(y_te, pred_gbt)
    
    # 모델 상황 확인
    
    print("분류기 개별 max_depth: ", 2 )
    print("분류기 개수: ", i)
    print("모델 정확도: ", accuracy)
    print("모델 분류 리포트")
    print(class_report)
    print("========================================================")

/Users/Angela/opt/anaconda3/envs/py3_8_5/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/Angela/opt/anaconda3/envs/py3_8_5/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/Users/Angela/opt/anaconda3/envs/py3_8_5/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, m

분류기 개별 max_depth:  2
분류기 개수:  10
모델 정확도:  0.6293706293706294
모델 분류 리포트
              precision    recall  f1-score   support

           0       0.00      0.00      0.00        53
           1       0.63      1.00      0.77        90

    accuracy                           0.63       143
   macro avg       0.31      0.50      0.39       143
weighted avg       0.40      0.63      0.49       143

분류기 개별 max_depth:  2
분류기 개수:  25
모델 정확도:  0.9370629370629371
모델 분류 리포트
              precision    recall  f1-score   support

           0       1.00      0.83      0.91        53
           1       0.91      1.00      0.95        90

    accuracy                           0.94       143
   macro avg       0.95      0.92      0.93       143
weighted avg       0.94      0.94      0.94       143

분류기 개별 max_depth:  2
분류기 개수:  50
모델 정확도:  0.951048951048951
모델 분류 리포트
              precision    recall  f1-score   support

           0       0.98      0.89      0.93        53
           1       0.94  

### 결과
```
분류기 개수:  50 / 
모델 정확도:  0.951048951048951

분류기 개수:  75 / 
모델 정확도:  0.958041958041958

분류기 개수:  100 /
모델 정확도:  0.965034965034965

분류기 개수:  150 / 
모델 정확도:  0.951048951048951
```
넣었던 개수 리스트 중 주요 변곡점을 짚어보면 이러하다. 100개일 때 가장 높은 성능을 보여주었고, 120개를 넘어 150개로 늘어났을 때의 정확도는 50개의 수준으로 다시 떨어졌다.

분류기 개수가 늘어날수록 성능이 올라가는 것은 아님을 이전 앙상블 모델들을 공부하면서도 보아왔다. 

그러나 그래디언트 부스팅의 개념을 생각해 보았을 때, 이전 학습기의 에러를 두고 공부를 하기 때문에 일정 수준의 성능까지 꾸준히 올라가 일정수준에서 고정될 것이라고 생각했으나, 그 것과는 다른 결과가 나온 것이다.

흥미로운 결과이다.
